govcd/metadata_v2.go

```diff
@@ -848,23 +869,156 @@ func mergeMetadataAndWait(client *Client, requestUri string, metadata map[string
 
 // deleteMetadata deletes metadata associated to the input key from an entity referenced by its URI, then returns the
 // task.
-func deleteMetadata(client *Client, requestUri string, key string, isSystem bool) (Task, error) {
+func deleteMetadata(client *Client, requestUri, name, key string, isSystem bool) (Task, error) {
 	apiEndpoint := urlParseRequestURI(requestUri)
 	if isSystem {
 		apiEndpoint.Path += "/metadata/SYSTEM/" + key
 	} else {
 		apiEndpoint.Path += "/metadata/" + key
 	}
 
+	err := filterMetadataToDelete(client, key, requestUri, name, isSystem, client.IgnoredMetadata)
+	if err != nil {
+		return Task{}, err
+	}
+
 	return client.ExecuteTaskRequest(apiEndpoint.String(), http.MethodDelete, "", "error deleting metadata: %s", nil)
 }
 
 // deleteMetadata deletes metadata associated to the input key from an entity referenced by its URI.
-func deleteMetadataAndWait(client *Client, requestUri string, key string, isSystem bool) error {
-	task, err := deleteMetadata(client, requestUri, key, isSystem)
+func deleteMetadataAndWait(client *Client, requestUri, name, key string, isSystem bool) error {
+	task, err := deleteMetadata(client, requestUri, name, key, isSystem)
 	if err != nil {
 		return err
 	}
 
 	return task.WaitTaskCompletion()
 }
+
+// IgnoredMetadata is a structure that defines the metadata entries that should be ignored by the VCD Client.
+// The filtering works in such a way that all the non-nil pointers in an instance of this struct are evaluated with
+// a logical AND.
+// For example, ignoredMetadata.ObjectType = "org", ignoredMetadata.ObjectName = "foo" will ignore all metadata from
+// Organizations whose name is "foo", with any key and any value.
+// Note: This struct is only used by metadata_v2.go methods.
+// Note 2: Filtering by ObjectName is not possible in the "ByHref" methods from VCDClient.
+type IgnoredMetadata struct {
+	ObjectType *string        // Type of the object that has the metadata as defined in the API documentation https://developer.vmware.com/apis/1601/vmware-cloud-director, for example "catalog", "disk", "org"...
+	ObjectName *string        // Name of the object
+	KeyRegex   *regexp.Regexp // A regular expression to filter out metadata keys
+	ValueRegex *regexp.Regexp // A regular expression to filter out metadata values
+}
+
+func (im IgnoredMetadata) String() string {
+	return fmt.Sprintf("IgnoredMetadata(ObjectType=%v, ObjectName=%v, KeyRegex=%v, ValueRegex=%v)", im.ObjectType, im.ObjectName, im.KeyRegex, im.ValueRegex)
```

Comment:
This needs a bit of improvement to print `ObjectType` and `ObjectName` values instead of only addresses as it won't give much value when troubleshooting.

```
2023/06/28 12:23:36 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=0x1400031a410, ObjectName=<nil>, KeyRegex=^foo$, ValueRegex=<nil>)'
2023/06/28 12:23:37 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=0x1400031a410, ObjectName=<nil>, KeyRegex=<nil>, ValueRegex=^bar$)'
2023/06/28 12:23:38 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=0x1400031a410, ObjectName=<nil>, KeyRegex=^fizz$, ValueRegex=<nil>)'
2023/06/28 12:23:38 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=0x1400031a410, ObjectName=<nil>, KeyRegex=<nil>, ValueRegex=^buzz$)'
2023/06/28 12:23:39 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=0x1400031a420, KeyRegex=^foo$, ValueRegex=<nil>)'
2023/06/28 12:23:39 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=0x1400031a420, KeyRegex=<nil>, ValueRegex=^bar$)'
2023/06/28 12:23:39 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=0x1400031a420, KeyRegex=^fizz$, ValueRegex=<nil>)'
2023/06/28 12:23:40 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=0x1400031a420, KeyRegex=<nil>, ValueRegex=^buzz$)'
2023/06/28 12:23:41 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=0x1400031a410, ObjectName=0x1400031a420, KeyRegex=foo, ValueRegex=bar)'
2023/06/28 12:23:41 [DEBUG] Comparing metadata with key 'test' with ignored metadata filter 'IgnoredMetadata(ObjectType=0x1400031a410, ObjectName=0x1400031a420, KeyRegex=foo, ValueRegex=bar)'
2023/06/28 12:23:47 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=<nil>, KeyRegex=^fo[a-z]$, ValueRegex=<nil>)'
2023/06/28 12:23:47 [DEBUG] Comparing metadata with key 'test' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=<nil>, KeyRegex=^fo[a-z]$, ValueRegex=<nil>)'
2023/06/28 12:23:47 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=<nil>, KeyRegex=^fo[a-z]$, ValueRegex=<nil>)'
2023/06/28 12:23:47 [DEBUG] Comparing metadata with key 'test' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=<nil>, KeyRegex=^fo[a-z]$, ValueRegex=<nil>)'
2023/06/28 12:23:47 [DEBUG] Comparing metadata with key 'foo' with ignored metadata filter 'IgnoredMetadata(ObjectType=<nil>, ObjectName=<nil>, KeyRegex=^fo[a-z]$, ValueRegex=<nil>)'
```

As I put you in this problem, I have investigated and it looks like pointers have to be handled manually. Made an experiment - https://go.dev/play/p/nDh-u4kJKoU

Reply:
Oops, this is definitely not what I wanted. Let me revisit it with your snippet.

Reply:
Fixed